keystoremanager/keystoremanager.go

```diff
@@ -155,79 +162,127 @@ attempts to validate the certificate by first checking for an exact match on
+	for _, keyID := range rootSigned.Signed.Roles["root"].KeyIDs {
+		// Decode all the x509 certificates that were bundled with this
+		// Specific root key
+		decodedCerts, err := trustmanager.LoadCertBundleFromPEM([]byte(rootSigned.Signed.Keys[keyID].Public()))
 		if err != nil {
 			logrus.Debugf("error while parsing root certificate with keyID: %s, %v", keyID, err)
 			continue
 		}
-		// TODO(diogo): Assuming that first certificate is the leaf-cert. Need to
-		// iterate over all decodedCerts and find a non-CA one (should be the last).
-		leafCert := decodedCerts[0]
 
+		// Get all non-CA certificates in the decoded certificates
+		leafCerts := trustmanager.GetLeafCerts(decodedCerts)
+
+		// If we got no leaf certificates or we got more than one, fail
+		if len(leafCerts) != 1 {
+			logrus.Debugf("error while parsing root certificate with keyID: %s, %v", keyID, err)
+			continue
+		}
+
+		// Get the ID of the leaf certificate
+		leafCert := leafCerts[0]
 		leafID, err := trustmanager.FingerprintCert(leafCert)
 		if err != nil {
 			logrus.Debugf("error while fingerprinting root certificate with keyID: %s, %v", keyID, err)
 			continue
 		}
 
-		// Check to see if there is an exact match of this certificate.
-		// Checking the CommonName is not required since ID is calculated over
-		// Cert.Raw. It's included to prevent breaking logic with changes of how the
-		// ID gets computed.
-		_, err = km.certificateStore.GetCertificateByKeyID(leafID)
-		if err == nil && leafCert.Subject.CommonName == dnsName {
-			certs[keyID] = rootSigned.Keys[keyID]
+		// Add all the valid leafs to the certificates map so we can refer to them later
+		allCerts[leafID] = leafCert
+
+		// Retrieve all the trusted certificates that match this dns Name
+		certsForCN, err := km.certificateStore.GetCertificatesByCN(dnsName)
+
+		// If there are no certificates with this CN, lets TOFUS!
+		// Note that this logic should only exist in docker 1.8
+		if len(certsForCN) == 0 {
+			km.certificateStore.AddCert(leafCert)
+			certsForCN = append(certsForCN, leafCert)
+			logrus.Debugf("using TOFUS on %s with keyID: %s", dnsName, leafID)
+		}
+
+		// Iterate over all known certificates for this CN and see if any are trusted
+		for _, cert := range certsForCN {
+			// Check to see if there is an exact match of this certificate.
+			certID, err := trustmanager.FingerprintCert(cert)
+			if err == nil && certID == leafID {
+				validKeys[keyID] = rootSigned.Signed.Keys[keyID]
+				logrus.Debugf("found an exact match for %s with keyID: %s", dnsName, keyID)
+			}
 		}
 
-		// Check to see if this leafCertificate has a chain to one of the Root CAs
-		// of our CA Store.
-		certList := []*x509.Certificate{leafCert}
-		err = trustmanager.Verify(km.caStore, dnsName, certList)
+		// Check to see if this leafCertificate has a chain to one of the Root
+		// CAs of our CA Store.
+		err = trustmanager.Verify(km.caStore, dnsName, decodedCerts)
```

Comment:
Can we skip this if we did the TOFUS thing instead? Or maybe it's better to look for a chain to a root CA before TOFUS...